src/main/ipc_exit.js

```diff
@@ -0,0 +1,89 @@
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { app, ipcMain, BrowserWindow } from "electron";
+import { removeUsbEventListeners } from "./ipc_device_discovery";
+
+export const registerExitHandlers = () => {
+  let readyToQuit = false;
+  let quitTimer = null;
+
+  ipcMain.on("app.exit", (event, arg) => {
+    app.quit();
+  });
+
+  // quit application when all windows are closed
+  app.on("window-all-closed", () => {
+    removeUsbEventListeners();
+
+    // on macOS it is common for applications to stay open until the user explicitly quits
+    if (process.platform !== "darwin") {
+      app.quit();
+    }
+  });
+
+  /*
+   * Our quit procedure is a bit involved, because if we have unsaved changes in
+   * the renderer, we want to offer a chance for the user to save those. So when
+   * we receive a "before-quit" event (initiated either from the main or the
+   * renderer process, or even from outside), we check if we have received a
+   * confirm-quit yet (readyToQuit will be true then, it's false by default). If
+   * we did, then we return early, and let the normal shutdown process commence.
+   *
+   * If we did not, then we set an 1s timer that'll automatically flip
+   * readyToQuit to true, and try to shut us down again. We do this so that if
+   * the renderer is unable to function correctly (because it crashed, or is
+   * showing an error screen or something), we can still quit.
+   *
+   * Once we have a timer rolling, we send an `app.should-quit?` message to all
+   * windows. They can, in turn, do an IPC call back: either `app.confirm-quit`
+   * to quit immediately, or `app.stop-quit` to prevent the process. Both of
+   * them clear the timer. The first sets readyToQuit and restarts the process,
+   * which will now quit immediately. The latter just cancels the timer, and
+   * then it's up to the application to handle the process further: ask the
+   * user, and then initiate another shutdown as necessary.
+   */
+
+  app.on("before-quit", (e) => {
+    // If we're already ready to quit, proceeed.
+    if (readyToQuit) return;
+
+    if (quitTimer) {
+      clearTimeout(quitTimer);
+    }
+    quitTimer = setTimeout(() => {
+      readyToQuit = true;
+      app.quit();
+    }, 1000);
+
+    // Before we quit, ask the renderer if we're ready.
+    BrowserWindow.getAllWindows().forEach((win) => {
+      win.webContents.send("app.should-quit?");
```

Comment:
Let's not stick questionmarks in channel names.

Reply:
Elsewhere we just do 'win.send', not win.webcontents.send, which suggests to me that one codepath or the other is wrong and *also* that we want a helper function so we alway do the same thing.

Reply:
According to the electron docs, to send a message from the main process to the renderer, we need `webContents.send()`. `BrowserWindow` - according to the docs - doesn't even have a `.send` method, so those are probably wrong.

I'll update the channel name, and add a helper.